crates/pallet-domains/src/staking.rs

```diff
@@ -272,20 +262,108 @@ pub(crate) fn do_deregister_operator<T: Config>(
     })
 }
 
+pub(crate) fn do_withdraw_stake<T: Config>(
+    operator_id: OperatorId,
+    nominator_id: NominatorId<T>,
+    withdraw: Withdraw<BalanceOf<T>>,
+) -> Result<(), Error> {
+    OperatorPools::<T>::try_mutate(operator_id, |maybe_operator_pool| {
+        let operator_pool = maybe_operator_pool.as_mut().ok_or(Error::UnknownOperator)?;
+        ensure!(!operator_pool.is_frozen, Error::OperatorPoolFrozen);
+
+        let nominator = Nominators::<T>::get(operator_id, nominator_id.clone())
+            .ok_or(Error::UnknownNominator)?;
+
+        let operator_owner =
+            OperatorIdOwner::<T>::get(operator_id).ok_or(Error::UnknownOperator)?;
+
+        let withdraw = match PendingWithdrawals::<T>::get(operator_id, nominator_id.clone()) {
+            None => withdraw,
+            Some(existing_withdraw) => match (existing_withdraw, withdraw) {
+                (Withdraw::All, _) => {
+                    // there is an existing full withdraw, error out
+                    return Err(Error::ExistingFullWithdraw);
+                }
+                (_, Withdraw::All) => {
+                    // there is exisiting withdrawal with specific amount,
+                    // since the new intent is complete withdrawl, use this instead
+                    Withdraw::All
+                }
+                (Withdraw::Some(previous_withdraw), Withdraw::Some(new_withdraw)) => {
+                    // combine both withdrawls into single one
+                    Withdraw::Some(
+                        previous_withdraw
+                            .checked_add(&new_withdraw)
+                            .ok_or(Error::BalanceOverflow)?,
+                    )
+                }
+            },
+        };
+
+        match withdraw {
+            Withdraw::All => {
+                // if nominator is the operator pool owner and trying to withdraw all, then error out
+                if operator_owner == nominator_id {
+                    return Err(Error::MinimumOperatorStake);
+                }
+
+                PendingWithdrawals::<T>::insert(operator_id, nominator_id, withdraw);
+            }
+            Withdraw::Some(withdraw_amount) => {
+                let total_pool_stake = operator_pool
+                    .current_total_stake
+                    .checked_add(&operator_pool.current_epoch_rewards)
+                    .ok_or(Error::BalanceOverflow)?;
+
+                let nominator_share =
+                    Perbill::from_rational(nominator.shares, operator_pool.total_shares);
+
+                let nominator_staked_amount = nominator_share * total_pool_stake;
```

Comment:
Should be rounding down `nominator_share.mul_floor(total_pool_stake)`

Reply:
I thought this was default behavior. Will change it to `mul_floor` instead